crates/test-programs/src/bin/p3_sockets_tcp_sample_application.rs

```diff
@@ -47,6 +47,7 @@ async fn test_tcp_sample_application(family: IpAddressFamily, bind_address: IpSo
 
             // Check that we sent and received our message!
             assert_eq!(data, first_message); // Not guaranteed to work but should work in practice.
+            drop(data_rx);
```

Comment:
I think this is an important detail of the implementation to call out - in the absence of task, the stream handle returned in `tuple<stream<T>, future<result<_, E>>>` pattern has to be dropped/closed before a value can arrive on the `future`, I suspect that is because in case of a task, the host could independently track the status of the socket and be notified of the socket shutdown by the OS as part of it's own event loop. Without a task, the guest must effectively "signal" the host to "do work", in success case triggering the `Drop` of the producer, I suppose.

I wonder if that's the expected behavior here and whether that would play well with GC languages.

Reply:
> Giving it a bit more thought, the guest is getting `StreamResult::Complete` on L46, shouldn't the host eagerly drop the producer _before_ letting the guest know that the stream is done?

Yes, it should; if it's not, I'd consider that a bug.  I'll take a look.